### PR TITLE
Migrates report filters

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Migrates and refactors report filters. 
Some other scattered fixes and todos. 

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
hey @latin-panda , would you mind having a look? Thanks for your time!

File: webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
+import { Component, EventEmitter, OnDestroy, ChangeDetectorRef, Output, ViewChild, Input } from '@angular/core';
+import { select, Store } from '@ngrx/store';
+import { Selectors } from '../../../selectors';
+import { combineLatest, Subscription, from } from 'rxjs';
+import { GlobalActions } from '../../../actions/global';
```
Comment (latin-panda) [inline webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:5]:
```suggestion
import { GlobalActions } from '@mm-actions/global';
```

File: webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import { MultiDropdownFilterComponent } from '@mm-components/filters/multi-dropdown-filter/mullti-dropdown-filter.component';
+import { PlaceHierarchyService } from '../../../services/place-hierarchy.service';
```
Comment (latin-panda) [inline webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:7]:
```suggestion
import { PlaceHierarchyService } from '@mm-services/place-hierarchy.service';
```

**Review by latin-panda — COMMENTED** [review]
This PR sparks joy ✨✨ I like the filters!!
I left few tiny suggestions. But in general everything looks good!
Would you like to include unit test or in the next PR?

File: webapp/src/ts/components/filters/date-filter/date-filter.component.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  @Output() search: EventEmitter<any> = new EventEmitter();
+
+  constructor(
+    private store: Store,
+  ) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  applyFilter() {
+    this.globalActions.setFilter({ date: this.date });
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/components/filters/date-filter/date-filter.component.ts:37]: You know what, I'm not sure why the filters are global. If it were up to me, I'd have had filters be dedicated. 
Maybe when this was added, there was something that needed access to filters in general? Maybe there still is? 

I guess we'll find out at the end of the migration :)

File: webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    this.globalActions.setFilter({ facilities: { selected: facilities.map(facility => facility.doc._id) } });
+    this.search.emit();
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  trackByFn(idx, facility) {
+    return facility.doc._id;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:97]: I changed it so it null checks the doc, thanks. 
We're not listening to changes here, this list doesn't "react" to changes, so we won't be redrawing items. But I can add the rev as well.

File: webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    return this.translateService.get(this.isAdmin ? 'place.deleted' : 'place.unavailable');
+  }
+
+  clear() {
+    this.dropdownFilter?.clear(false);
+  }
+}
+
+
+/*angular.module('inboxDirectives').directive('mmFacilityFilter', function(SearchFilters) {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:119]: It's not complete, there is one todo left which is a performance issue. 
I think it's too early to test the performance of any part of the application, so I'm leaving it as a todo for now and keeping this comment in so I remember.

File: webapp/src/ts/components/filters/form-type-filter/form-type-filter.component.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  private globalActions;
+  forms;
+
+  @Input() disabled;
+  @Output() search: EventEmitter<any> = new EventEmitter();
+  @ViewChild(MultiDropdownFilterComponent)
+  dropdownFilter: MultiDropdownFilterComponent;
+
+  constructor(
+    private store:Store,
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/components/filters/form-type-filter/form-type-filter.component.ts:25]: Is this the stylistic approach? 

I find it really hard to differentiate between Object assigns and type declarations visually, so I keep the space for object property assignments `{ a: 'something' }` and no space when doing type declarations `{ a:string }`.

File: webapp/src/ts/modules/reports/reports-filters.component.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @Output() search: EventEmitter<any> = new EventEmitter();
+
+  @ViewChild(FormTypeFilterComponent)
+  formTypeFilter:FormTypeFilterComponent;
+  @ViewChild(FacilityFilterComponent)
+  facilityFilter:FacilityFilterComponent;
+  @ViewChild(DateFilterComponent)
+  dateFilter:DateFilterComponent;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/reports/reports-filters.component.ts:32]: They are generic (and were pretty generic before too). 
We reuse the freetext filter in contacts.

File: webapp/src/ts/modules/reports/reports.component.ts [hunk 8]
```diff
@@ -191,10 +209,30 @@ export class ReportsComponent implements OnInit, OnDestroy {
...
     return this.query(force);
   }
+
+  resetFilterModel() {
+    if (this.selectMode && this.selectedReports && this.selectedReports.length) {
+      // can't filter when in select mode
+      return;
+    }
+    this.globalActions.clearFilters();
+    // todo reset filters!
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/reports/reports.component.ts:233]: Thanks for pointing this out! Nope, not needed anymore.

**Comment by dianabarsan** [comment]
hey @latin-panda 

I've added/migrated unit tests and changed some minor things as a result. :D 
I also had an attempt to fixing the unit test build job, and it was one thing to fix after another :)) 
But I fixed it, the stupid build works! https://travis-ci.org/github/medic/cht-core/jobs/731598931

anyway, please have another look. Thanks for your time!

File: webapp/src/ts/modules/reports/reports.component.ts [hunk 9]
```diff
@@ -191,7 +209,17 @@ export class ReportsComponent implements OnInit, OnDestroy {
   };
 
   search(force = false) {
-    // todo filters
+    // clears report selection for any text search or filter selection
+    // does not clear selection when someone is editing a form
+    if((this.filters.search || Object.keys(this.filters).length > 1) && !this.enketoEdited) {
+      //$state.go('reports.detail', { id: null }, { notify: false });
```
Comment (latin-panda) [inline webapp/src/ts/modules/reports/reports.component.ts:215]: Do we still need this "if"?

**Review by latin-panda — APPROVED** [review]
Excellent very neat!! I don't really have additional comments, just added 1 question inline and I have this other question:

Are the following files going to be migrated/finalised in future PRs? Just wanted to double check that these weren't accidentally forgotten
/tests/karma/ts/pipes/date.spec.js 
/tests/karma/ts/providers/lineage.spec.js  
/tests/karma/ts/reducers/reports.spec.ts

File: webapp/src/ts/modules/reports/reports.component.ts (hunk 9 above)
Comment (dianabarsan) [reply webapp/src/ts/modules/reports/reports.component.ts:215]: Yea, I think so. It's still commented because we don't have a report detail page yet. :D

**Comment by dianabarsan** [comment]
> Are the following files going to be migrated/finalised in future PRs? Just wanted to double check that these weren't accidentally forgotten
/tests/karma/ts/pipes/date.spec.js
/tests/karma/ts/providers/lineage.spec.js
/tests/karma/ts/reducers/reports.spec.ts

Not forgotten, but I wanted to get this PR over with. I should have added those in the initial branch but tests weren't setup yet :( 
My plan is to: get the detail page working and then migrate all the skipped unit tests (there are a bunch!!)

**Pull request merged**